src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.CopyFile.OSX.cs

```diff
@@ -0,0 +1,123 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.IO
+{
+    internal static partial class FileSystem
+    {
+        public static partial void CopyFile(string sourceFullPath, string destFullPath, bool overwrite)
+        {
+            // Fail fast for blatantly copying onto self
+            if (sourceFullPath == destFullPath)
+            {
+                throw new IOException(SR.Format(overwrite ? SR.IO_SharingViolation_File : SR.IO_FileExists_Name, destFullPath));
+            }
+
+            // Start by locking, creating relevant file handles, and reading out file status info.
+            using SafeFileHandle src = SafeFileHandle.OpenReadOnly(sourceFullPath, FileOptions.None, out Interop.Sys.FileStatus fileStatus);
+            UnixFileMode filePermissions = SafeFileHandle.GetFileMode(fileStatus);
+
+            // Read FileStatus of destination file to determine how to continue
+            int destError = Interop.Sys.Stat(destFullPath, out Interop.Sys.FileStatus destStat);
+
+            // Interpret the error from stat
+            if (destError != 0)
+            {
+                // Some error, let's see what it is:
+                Interop.Error error = Interop.Sys.GetLastError();
+
+                // Destination not existing is expected, so if this is the case we try clonefile,
```

Comment:
I assume you mean destination instead of directory since that's what is says in the comment. If the destination is created before `clonefile` creates the file, it will return an error of `EEXIST`, which is handled in the `tryCloneFile` section to go to the fallback (unless we're not overriding, in which case it just throws the error then). The reason we need this check is because `clonefile` always fails if the destination exists, which is also the reason we have the deleting logic.
I will update the comment to say `in this case we try clonefile now since the destination file doesn't exist` or something similar.

Reply:
The code is designed so that if pretty much anything unexpected happens that we don't explicitly handle, we should end up on the fallback path. Should I add a comment stating that, since people may want to change the code at some point.

Reply:
I'm asking why we need any of this:
https://github.com/dotnet/runtime/pull/79243/files#diff-26c8966a5f0a0f29694825424de6d26bec54a7f81b3cd6e6fc03dc08e5c0ad0dR24-R73
Between the time we check whether the destination file exists (and find that it doesn't exist) and the time that we call clonefile, the destination file could be created.  As such the code for clonefile needs to be able to deal with that.  And if what it's currently doing to deal with that is correct, then why do we need any of the stuff before it?  We can just use clonefile for the case it's meant to handle: when the destination doesn't exist.  If someone explicitly specifies overwrite:true and the destination file exists, clonefile will fail and we'll just fall through to the normal copy path.

Reply:
> We can just use clonefile for the case it's meant to handle: when the destination doesn't exist. If someone explicitly specifies overwrite:true and the destination file exists, clonefile will fail and we'll just fall through to the normal copy path.

That would make the optimisation inapplicable half of the time for no reason.
I feel like if we did this (which I would be against) that it must be documented, and I personally don't want to add to the documentation "to make `File.Copy` faster on macOS, you should delete the destination manually first since we specifically decided not to." (it would obviously not be worded quite like that)

> Between the time we check whether the destination file exists (and find that it doesn't exist) and the time that we call clonefile, the destination file could be created.

This is unlikely and would be caught by the fallback path anyway.

> As such the code for clonefile needs to be able to deal with that.

I initially had code that retried doing `clonefile` when it failed due to the destination came into existence, but someone asked me to remove it (I don't recall who, but it could be found if you could be bothered) because it was overcomplicating it for an edge case.

> And if what it's currently doing to deal with that is correct, then why do we need any of the stuff before it?

The stuff before it does multiple things to check if it's worth trying to use `clonefile` and to make `clonefile` work in the case of the destination existing.

Specifically,
1. If the destination to exist then we need to delete it, this is checked using `stat`.
2. If it's on a different device `clonefile` won't work (we get this info for free from `stat`).
3. If we're copying onto ourselves in a non-obvious way (via symlinks), then `File.Copy` should fail, there is no error for this with `clonefile`. (i.e. we need to call `stat` on the destination anyway since `File.Copy` should throw an exception in this case, even if we didn't care about checking the existence of the destination)
4. The deleting section is self explanatory since `clonefile` will fail if the destination exists.

We get all of the other information for free when determining if the destination exists & when checking if we're copying onto ourselves by using `stat`, so it would be a waste to not use it to short circuit particular cases.

Let me know if any of that doesn't make sense.

Reply:
I agree with @hamarb123 here. The original use case for the optimization was MSBuild which does copying of the files a lot. In majority of the cases a rebuild is happening and the destination files exist. That's also the reason to introduce the path with deleting the destination file and only falling back to standard copy on any subsequent failure.

Reply:
The code should start by making a `clonefile` call.

If the error is `EEXIST` and the arg `overwrite` then do the delete and call `clonefile` once more.

For the delete, I don't think we need to lock the `dst` (it doesn't bring much), and rather than using `File.Delete` use `Unlink` so we don't throw something unexpected.

Reply:
@tmds The locking is needed, and there was an issue raised about it. It is observable by user code - https://github.com/dotnet/runtime/issues/31429. I don't think there's point in calling `clonefile` first since the file locking check has to be made and we already have all the information about the destination file, and thus we know the `clonefile` call would fail.

Reply:
> For the delete, I don't think we need to lock the dst (it doesn't bring much)

@filipnavara may remember, but iirc there was complaints when this was done before.

I personally would be more than happy for this whole locking nonsense to go away since it doesn't really exist on Unix how it does, it would also greatly simplify the normal Unix implementation, but I feel like that would be a major breaking change since apparently people rely on it.

And if we have to check this properly, then we may as well have the rest of it.

Reply:
> Calling `clonefile` first is a way to eliminate the `stat` syscall.

We need to throw when we copy onto ourselves unless we want to make a breaking change, I don't believe `copyfile` has any issues with doing this from looking at the documentation.

Reply:
Actually, we should be able to attempt to open a lock on the destination file, which should fail with the appropriate exception if we're copying onto ourselves.

Reply:
> That would make the optimisation inapplicable half of the time for no reason.

I wouldn't be pushing on it is there were "no reason". There are multiple reasons.
1. This is much more complex. Complexity comes with bugs, maintainability burden, compatibility burden, etc.
2. Doing a stat incurs an expensive syscall for everyone regardless of whether clonefile will be successful.
3. Deleting the file is irreversible. If we delete the destination and then for example we fail to open the source again, the copy has failed and the destination is gone. That's data loss.
4. At this level of the stack we don't have the luxury of saying "oh well, if there's a race condition our behavior doesn't matter". These kinds of races absolutely happen. I've seen many occurrences for example of a File.Exists check returning true and then a File.Open immediately after throwing because the file is then gone.

I'll catch up on the rest of the discussion that happened over my night...

Reply:
> Deleting the file is irreversible. If we delete the destination and then for example we fail to open the source again, the copy has failed and the destination is gone. That's data loss.

There's an open handle for the source file that prevents this specific case from happening. I would not necessarily be worried about data loss when the user explicitly indicated with `overwrite: true` that the destination file can be overwritten. While not identical, it's similar to the fallback code path truncating the destination file and then failing to write into it (eg. because the disk is full).

Reply:
> I wouldn't be pushing on it if there were "no reason".

Sorry that I worded it like that, that's not how I meant it when I wrote it, but I can see how I've caused this miscommunication.

> Deleting the file is irreversible. If we delete the destination and then for example we fail to open the source again, the copy has failed and the destination is gone. That's data loss.

@filipnavara summed up my thoughts on this pretty well, and worded it better than I would have :)

> Doing a stat incurs an expensive syscall for everyone regardless of whether clonefile will be successful.

If we want to support overwriting the destination file and still using `clonefile`, then I don't see how we could avoid having either an existence check (by way of opening the file handle), or a stat check, assuming we want to keep the locking logic. If we did the existence check first, I think it is likely that the code would look much more complex.

I would also note that while a stat call being expensive is true, I would think that a full file copy is far more expensive than the stat check anyway (I would think this is true, even for small files, since it requires writing to disk rather than just reading, and most of the info from stat would also be needed as a part of that), thus this should be a net positive in the general case.

> This is much more complex. Complexity comes with bugs, maintainability burden, compatibility burden, etc.

This is a fair criticism, but I would just like to mention that this complexity is almost solely caused by the complex locking logic that we have decided to maintain.

I'm also happy to go and add comments explaining why things are done the way they are to achieve certain guarantees, but I feel like (based on some of your previous feedback that seemed to be effectively saying that there were too many comments) that they would just be requested to be removed. If I'm wrong and you would like me to add these comments, please let me know.

Reply:
Something like this:
```cs
if (Interop.@libc.clonefile(sourceFullPath, destFullPath, Interop.@libc.CLONE_ACL) == 0)
{
    return true;
}
Interop.Error error = Interop.Sys.GetLastError();

// If the file exists and we're overwriting, delete it and call clonefile again.
if (error == Interop.Error.EEXIST && overwrite)
{
    ... lock, check same, delete ...

    if (Interop.@libc.clonefile(sourceFullPath, destFullPath, Interop.@libc.CLONE_ACL) == 0)
    {
        return true;
    }
    error = Interop.Sys.GetLastError();

    if (error == Interop.Error.EEXIST)
    {
        return false; // (unlikely) fall back to regular copy.
    }
}

if (error == Interop.Error.ENOTSUP ||
    error == Interop.Error.EXDEV)
{
    return false; // fall back to regular copy.
}

Debug.Assert(error != Interop.Error.SUCCESS);
Interop.CheckIo(error);
```

Reply:
@tmds, let me think about it a little more to ensure there isn't any issues (I'm also a bit busy this week :) ), but otherwise I'll do that. Will implement your other suggestions along with it also.

Reply:
@tmds, this has the issue of allowing to copy onto ourselves. Do we want that to be allowed to make it work?

Reply:
I've realised that wouldn't matter since if we're copying onto ourselves, then the destination presumably exists, so it should be fine. Will implement it soon

Reply:
> if we're copying onto ourselves, then the destination presumably exists,

This test will verify the behavior:

https://github.com/dotnet/runtime/blob/c79d74897b8501f2c0fea1e59131ce2b34a1b702/src/libraries/System.IO.FileSystem/tests/File/Copy.cs#L44

Reply:
That wouldn't test for via a symlink, but if we're not testing that the name of the paths are the same, then it should be a fine way to test it.